pyquil/quil.py

```diff
@@ -719,15 +719,59 @@ def instantiate_labels(instructions):
     return result
 
 
+def merge_with_pauli_noise(prog_list, probabilities, qubits):
+    """
+    Insert pauli noise channels between each item in the list of programs.
+    This noise channel is implemented as a single noisy gate acting on the provided qubits.
+
+    :param prog_list: an iterable such as a program or a list of programs.
+    If a program is provided, a single noise gate will be applied after each gate in the program.
+    If a list of programs is provided, the noise gate will be applied after each program.
+    :param list|floats probabilities: The 4^num_qubits list of probabilities specifying the desired pauli channel.
+    There should be either 4 or 16 probabilities specified in the order I,X,Y,Z or II, IX, IY, IZ, XI, XX, XY, etc
+    :param list qubits: a list of the qubits that the noisy gate should act on.
+    :return: A single program with noisy gates inserted between each element of the program list.
+    :rtype: Program
+    """
+    p = Program()
+    p.defgate("pauli_noise", np.eye(2 ** len(qubits)))
+    p.define_noisy_gate("pauli_noise", qubits, pauli_kraus_map(probabilities))
+    for elem in prog_list:
+        p.inst(Program(elem)).inst(("pauli_noise", *qubits))
+    return p
+
+
 def merge_programs(prog_list):
     """
-    Merges a list of pyQuil programs into a single one by appending them in sequence
+    Merges a list of pyQuil programs into a single one by appending them in sequence.
+    If multiple programs in the list contain the same gate and/or noisy gate definition
+    with identical name, this definition will only be applied once. If different definitions
+    with the same name appear multiple times in the program list, each will be applied once
+    in the order of last occurrence.
 
     :param list prog_list: A list of pyquil programs
     :return: a single pyQuil program
     :rtype: Program
     """
-    return sum(prog_list, Program())
+    definitions = [gate for prog in prog_list for gate in prog.defined_gates]
+    seen = {}
+    # Collect definitions in reverse order and reapply definitions in reverse
+    # collected order to ensure that the last occurrence of a definition is applied last.
+    for definition in reversed(definitions):
+        name = definition.name
+        if name in seen.keys():
+            if definition not in seen[name]:
+                seen[name] += [definition]
+        else:
+            seen[name] = [definition]
+    new_definitions = [gate for key in seen.keys() for gate in reversed(seen[key])]
```

Comment:
If you're relying on an ordering `seen`, it needs to be a list - dictionaries aren't guaranteed to be ordered.

Reply:
I'm relying on the order of the list stored in the dictionary entry seen[key]. I don't think I'm relying on any ordering in seen itself.

Reply:
Alright, so you don't mind if `new_definitions`'s order changes?